### PR TITLE
Migrate to a good Content Security Policy

### PR DESCRIPTION
**Pull request by leonklingele**
So attacks such as XSS (see [0]) will no longer be of an issue.

[0]: https://github.com/omarroth/invidious/issues/1022

__TODO__:

- [x] Adapt CSP
- [x] Migrate `onclick=..`, `onsubmit=..` etc. directives
- [x] Migrate inline JS `<script> [some JS] </script>` to its own files
- [x] Test!
- [ ] Optional: Migrate `javascript:` links, will log warnings but that can be ignored

### CODE REVIEW

File: assets/css/default.css [hunk 1]
```diff
@@ -447,6 +447,17 @@ video.video-js {
   height: 0;
 }
 
+#player {
```
Comment (leonklingele) [inline assets/css/default.css]: @omarroth is this still required? Moving to `default.css` seems to break the player layout, works fine without any styles applied to `#player`.

**Comment by leonklingele** [comment]
Feel free to test this now! :)

File: assets/js/silvermine-videojs-quality-selector.min.js [hunk 2]
```diff
@@ -1,3 +1,3 @@
-/*! @silvermine/videojs-quality-selector 2019-09-26 v1.2.2-4-gc134430-dirty */
```
Comment (omarroth) [inline assets/js/silvermine-videojs-quality-selector.min.js]: You'll want to use [this fork](https://github.com/omarroth/videojs-quality-selector), which has a fix for #158 and silvermine/videojs-quality-selector#57 applied. You can grab a pre-built copy [here](https://omar.yt/js/silvermine-videojs-quality-selector.min.js).

File: assets/css/default.css (hunk 1 above)
Comment (omarroth) [reply assets/css/default.css]: The `#player` styling is only really necessary when viewing the embed page with javascript disabled in order for the player to resize correctly.

Comment (omarroth) [reply assets/css/default.css]: I misspoke, it still appears to be necessary even javascript enabled.

File: assets/js/silvermine-videojs-quality-selector.min.js (hunk 2 above)
Comment (leonklingele) [reply assets/js/silvermine-videojs-quality-selector.min.js]: Updated.

File: assets/css/default.css (hunk 1 above)
Comment (leonklingele) [reply assets/css/default.css]: Updated

**Comment by leonklingele** [comment]
@omarroth ptal

File: assets/css/default.css [hunk 3]
```diff
@@ -5,6 +5,17 @@ body {
     Arial, sans-serif;
 }
 
+#player {
```
Comment (omarroth) [inline assets/css/default.css]: Sorry I didn't mention this correctly: this is only needed for the embed page. It should either be in `embed.ecr` or a separate CSS file.

File: src/invidious/views/add_playlist_items.ecr [hunk 4]
```diff
@@ -20,9 +20,9 @@
     <div class="pure-u-1 pure-u-lg-1-5"></div>
 </div>
 
-<script>
-var playlist_data = {
-    csrf_token: '<%= URI.encode_www_form(env.get?("csrf_token").try &.as(String) || "") %>',
+<script id="playlist_data" type="application/json">
+{
+    "csrf_token": "<%= URI.encode_www_form(env.get?("csrf_token").try &.as(String) || "") %>",
```
Comment (omarroth) [inline src/invidious/views/add_playlist_items.ecr]: There are a couple trailing commas that are valid javascript objects but invalid JSON. I've done my best to point out any I noticed.

File: src/invidious/views/community.ecr [hunk 5]
```diff
@@ -71,14 +71,14 @@
...
-    show_replies_text: '<%= HTML.escape(translate(locale, "Show replies")) %>',
-    preferences: <%= env.get("preferences").as(Preferences).to_json %>,
+<script id="community_data" type="application/json">
+{
+    "ucid": "<%= channel.ucid %>",
+    "youtube_comments_text": "<%= HTML.escape(translate(locale, "View YouTube comments")) %>",
+    "comments_text": "<%= HTML.escape(translate(locale, "View `x` comments", "{commentCount}")) %>",
+    "hide_replies_text": "<%= HTML.escape(translate(locale, "Hide replies")) %>",
+    "show_replies_text": "<%= HTML.escape(translate(locale, "Show replies")) %>",
+    "preferences": <%= env.get("preferences").as(Preferences).to_json %>,
```
Comment (omarroth) [inline src/invidious/views/community.ecr]: See above.

File: src/invidious/views/playlist.ecr [hunk 6]
```diff
@@ -69,9 +69,9 @@
 </div>
 
 <% if playlist.is_a?(InvidiousPlaylist) && playlist.author == user.try &.email %>
-<script>
-var playlist_data = {
-    csrf_token: '<%= URI.encode_www_form(env.get?("csrf_token").try &.as(String) || "") %>',
+<script id="playlist_data" type="application/json">
+{
+    "csrf_token": "<%= URI.encode_www_form(env.get?("csrf_token").try &.as(String) || "") %>",
```
Comment (omarroth) [inline src/invidious/views/playlist.ecr]: See above.

File: src/invidious/views/subscriptions.ecr [hunk 7]
```diff
@@ -45,9 +45,9 @@
     <hr>
 </div>
 
-<script>
-var watched_data = {
-    csrf_token: '<%= URI.encode_www_form(env.get?("csrf_token").try &.as(String) || "") %>',
+<script id="watched_data" type="application/json">
+{
+    "csrf_token": "<%= URI.encode_www_form(env.get?("csrf_token").try &.as(String) || "") %>",
```
Comment (omarroth) [inline src/invidious/views/subscriptions.ecr]: See above.

File: src/invidious/views/template.ecr [hunk 8]
```diff
@@ -147,13 +147,14 @@
...
     <% if env.get? "user" %>
         <script src="/js/sse.js?v=<%= ASSET_COMMIT %>"></script>
-        <script>
-            var notification_data = {
-                upload_text: '<%= HTML.escape(translate(locale, "`x` uploaded a video")) %>',
-                live_upload_text: '<%= HTML.escape(translate(locale, "`x` is live")) %>',
+        <script id="notification_data" type="application/json">
+            {
+                "upload_text": "<%= HTML.escape(translate(locale, "`x` uploaded a video")) %>",
+                "live_upload_text": "<%= HTML.escape(translate(locale, "`x` is live")) %>",
```
Comment (omarroth) [inline src/invidious/views/template.ecr]: See above.

File: src/invidious/views/history.ecr [hunk 9]
```diff
@@ -18,9 +18,9 @@
     </div>
 </div>
 
-<script>
-var watched_data = {
-    csrf_token: '<%= URI.encode_www_form(env.get?("csrf_token").try &.as(String) || "") %>',
+<script id="watched_data" type="application/json">
+{
+    "csrf_token": "<%= URI.encode_www_form(env.get?("csrf_token").try &.as(String) || "") %>",
```
Comment (omarroth) [inline src/invidious/views/history.ecr]: See above.

**Comment by leonklingele** [comment]
ptal @omarroth

**Comment by omarroth** [comment]
LGTM :+1:

File: assets/js/global.js [hunk 10]
```diff
@@ -0,0 +1,3 @@
+// Disable Web Workers. Fixes Video.js CSP violation (created by `new Worker(objURL)`):
+// Refused to create a worker from 'blob:http://host/id' because it violates the following Content Security Policy directive: "worker-src 'self'".
+window.Worker = undefined;
```
Comment (tenpura-shrimp) [inline assets/js/global.js:3]: excuse me what the fuck

**Pull request merged**